### ISSUE TITLE
Introduce dirty flags to limit expensive requests

### ISSUE THREAD

**Issue by sclu1034** [kak-lsp/kak-lsp#663]
kak-lsp requests inlay hints and semantic tokens on every `NormalIdle`, even when all I'm doing is move the cursor. Depending on the language server and file size, the responses for that can be quite big.
Additionally, when `NormalIdle` is triggered by some form of "menuing" (e.g. running a command like `lsp-hover-buffer` that doesn't affect the buffer at all), then not even the cursor changes, and reference highlighting doesn't need updates either.

If possible, there should be some form of "dirty" flags that track cursor movement and file changes and that prevent certain requests from running when they are not necessary, at least when triggered by `NormalIdle` hooks.

**Comment by krobelus** [comment]
yeah inlay hints and semantic tokens do not depend on the cursor position, so we should only re-send them if the buffer changed.
The existing code already tries to do that via `lsp-did-change` which does

```kak
    evaluate-commands %sh{
        if [ $kak_opt_lsp_timestamp -eq $kak_timestamp ]; then
            echo "fail"
        fi
    }
    set-option buffer lsp_timestamp %val{timestamp}
```

I don't understand how that even works right now. It reads to me like the first hook to run lsp-did-change makes the other hooks a nop.. but that can't be true

**Comment by sclu1034** [comment]
It does look like that. It's scoped by buffer instead of global, at least, but still.

On the other hand, evidence of all requests running always suggests that it somehow doesn't prevent anything.

**Comment by krobelus** [comment]
I'll look into it.

At least for semantic tokens we can switch to the refresh variant (making the server is responsible for sending updates) as proposed in https://github.com/kak-lsp/kak-lsp/pull/546/commits/aa9af5e220126313dea7f035a43a0cdd80a870f5

**Comment by krobelus** [comment]
> Additionally, when NormalIdle is triggered by some form of "menuing" (e.g. running a command like lsp-hover-buffer that doesn't affect the buffer at all),

for inlay hints and semantic highlighting, that's a clear improvement but for auto-hover we may want to still show it, I'm not sure

Here's a change that should do the trick.
Probably we should refactor it to do it in the same shell call as lsp-did-change.
Or do the cache logic on the server side..
In future we should reimplement all lsp commands by echoing data to a socket (like pykak), then we'll need almost no shell calls.

```diff
diff --git a/rc/lsp.kak b/rc/lsp.kak
index ae9f0b38..9b632037 100644
--- a/rc/lsp.kak
+++ b/rc/lsp.kak
@@ -1310,7 +1310,13 @@ incomingOrOutgoing = $1
 
 
 define-command -hidden lsp-inlay-hints -docstring "lsp-inlay-hints: request inlay hints" %{
-  lsp-did-change-and-then lsp-inlay-hints-request
+    declare-option -hidden str lsp_inlay_hints_cache_key
+    evaluate-commands %sh{
+        if [ "$kak_opt_lsp_inlay_hints_cache_key" != "$kak_timestamp" ]; then
+            echo set-option buffer lsp_inlay_hints_cache_key %val{timestamp}
+            echo lsp-did-change-and-then lsp-inlay-hints-request
+        fi
+    }
 }
 
 define-command -hidden lsp-inlay-hints-request %{
@@ -1560,7 +1566,13 @@ define-command tsserver-organize-imports -docstring "Ask typescript-language-ser
 # semantic tokens
 
 define-command lsp-semantic-tokens -docstring "lsp-semantic-tokens: Request semantic tokens" %{
-  lsp-did-change-and-then lsp-semantic-tokens-request
+    declare-option -hidden str lsp_semantic_tokens_cache_key
+    evaluate-commands %sh{
+        if [ "$kak_opt_lsp_semantic_tokens_cache_key" != "$kak_timestamp" ]; then
+            echo set-option buffer lsp_semantic_tokens_cache_key %val{timestamp}
+            echo lsp-did-change-and-then lsp-semantic-tokens-request
+        fi
+    }
 }
 
 define-command -hidden lsp-semantic-tokens-request %{
@@ -2029,10 +2041,18 @@ define-command lsp-auto-hover-enable -params 0..1 -client-completion \
     -docstring "lsp-auto-hover-enable [<client>]: enable auto-requesting hover info for current position
 
 If a client is given, show hover in a scratch buffer in that client instead of the info box" %{
+    declare-option -hidden str lsp_auto_hover_cache_key
     evaluate-commands %sh{
         hover=lsp-hover
         [ $# -eq 1 ] && hover="lsp-hover-buffer $1"
-        printf %s "hook -group lsp-auto-hover global NormalIdle .* %{ $hover }"
+        printf %s 'hook -group lsp-auto-hover global NormalIdle .* %{
+            evaluate-commands %sh{
+                if [ "$kak_opt_lsp_auto_hover_cache_key" != "$kak_timestamp $kak_selections_desc" ]; then
+                    echo set-option window lsp_auto_hover_cache_key \"%val{timestamp} %val{selections_desc}\"
+                    echo '"$hover"'
+                fi
+            }
+        }'
     }
 }
```

**Comment by krobelus** [comment]
so `lsp-did-change` always calls its callback, intentionally.

semantic tokens and inlay hints should be fixed by 9621c99d8ad358af806d4d88f43cd1e50bec87c7, thanks for reporting

I checked the other requests; most of them do not send a lot of data so it's
probably not worth caching them. Also most requests depend on the cursor
position, so caching is less effective.
Also depends on which behavior we want for auto-hover, do we want the hover info to pop back up after pressing escape etc..  I don't really use it.

So I think we're fine now.
